### PR TITLE
implement the category

### PR DESCRIPTION
**Pull request by volodymyrruso**
## Summary by CodeRabbit

- **New Features**
  - Introduced a RESTful API for category management, allowing creation, retrieval, updating, and deletion of categories.
  - Added functionality to fetch books by category.
  - Implemented a shopping cart management system, enabling users to add, update, and delete items in their shopping carts.
  - Introduced DTOs for cart items and shopping carts for better data handling.

- **Bug Fixes**
  - Enhanced data access patterns in the BookRepository for more sophisticated querying.

- **Documentation**
  - Updated database migration scripts to include new tables and columns for categories and their relationships with books, as well as for the shopping cart.

- **Chores**
  - Increased JWT expiration time for enhanced security token validity.

### CODE REVIEW

File: src/main/java/org/mate/bookstorespringboot/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PostMapping("/create")
+    @PreAuthorize("hasRole('ADMIN')")
+    public ResponseEntity<CategoryDto> createCategory(@Valid @RequestBody
+                                                          CategoryRequestDto categoryRequestDto) {
+        CategoryDto savedBook = categoryService.save(categoryRequestDto);
+        return ResponseEntity.status(HttpStatus.CREATED).body(savedBook);
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/controller/CategoryController.java]: You can use ResponseSytatus annotation

File: src/main/java/org/mate/bookstorespringboot/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    @PutMapping("{id}")
+    @PreAuthorize("hasRole('ADMIN')")
+    public ResponseEntity<CategoryDto> updateCategory(@PathVariable Long id,
+                                                      @RequestBody @Valid
+                                                      CategoryRequestDto categoryRequestDto) {
+        return ResponseEntity.ok(categoryService.updateCategory(id, categoryRequestDto));
+    }
+
+    @DeleteMapping("{id}")
+    @PreAuthorize("hasRole('ADMIN')")
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/controller/CategoryController.java:64]: use response status no content

File: src/main/java/org/mate/bookstorespringboot/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    public void deleteCategoryById(@PathVariable Long id) {
+        categoryService.deleteById(id);
+    }
+
+    @GetMapping("{id}/books")
+    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
+    public ResponseEntity<List<BookDtoWithoutCategoryIds>> getBookByCategoryId(
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/controller/CategoryController.java]: Better return Page instead List

File: src/main/java/org/mate/bookstorespringboot/controller/dto/book/BookDto.java [hunk 4]
```diff
@@ -1,7 +1,9 @@
 package org.mate.bookstorespringboot.controller.dto.book;
 
 import java.math.BigDecimal;
+import lombok.Builder;
 
+@Builder
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/controller/dto/book/BookDto.java]: Does it work with record class? https://medium.com/thefreshwrites/migrating-from-lombok-to-records-in-java-fa2ece28bd68

File: src/main/java/org/mate/bookstorespringboot/controller/mapper/BookMapper.java [hunk 5]
```diff
@@ -20,4 +22,11 @@ default Book toEntityWithId(BookRequestDto bookRequestDto, Long id) {
         book.setId(id);
         return book;
     }
+
+    BookDtoWithoutCategoryIds toDtoWithoutCategoryIds(Book book);
+
+    @AfterMapping
+    default void setCategoryId(Book book, BookDto bookDto) {
+
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/controller/mapper/BookMapper.java:34]: Should you add some realization?

File: src/main/java/org/mate/bookstorespringboot/repository/BookRepository.java [hunk 6]
```diff
@@ -1,9 +1,21 @@
...
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 
 @Repository
 public interface BookRepository extends JpaRepository<Book, Long> {
+
+    @EntityGraph(attributePaths = "categories")
+    Collection<Object> findAll(Specification<Book> bookSpecification);
+
+    @Query("FROM Book b JOIN b.categories c WHERE c.id = :categoryId")
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/repository/BookRepository.java]: do you really need a custom query here?

File: src/main/java/org/mate/bookstorespringboot/service/impl/BookServiceImpl.java [hunk 7]
```diff
@@ -55,6 +59,13 @@ public void deleteById(Long id) {
         bookRepository.deleteById(id);
     }
 
+    @Override
+    public List<BookDtoWithoutCategoryIds> findAllByCategoryId(Long id, Pageable pageable) {
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/service/impl/BookServiceImpl.java]: return Page here

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! A few recommendation ;)

File: src/main/java/org/mate/bookstorespringboot/controller/mapper/BookMapper.java (hunk 5 above)
Comment (volodymyrruso) [reply src/main/java/org/mate/bookstorespringboot/controller/mapper/BookMapper.java:30]: not really, forgot to deleted it before pushing 🫨

**Pull request merged**